### PR TITLE
feat(docs): fetch remote docs from GitHub

### PR DESCRIPTION
**Pull request by CodyJasonBennett**
Implements #86 by fetching docs from libraries' respective repositories instead of solely from this repo.

Some things to address:
 - [x] target configuration (opt-in via `data/libraries` under the `docs` key, configured to fetch from this repo by default)
 - [x] file-based routing with categories (read as `<category-name>/<page-name>.md(x)` and will defer to using MDX frontmatter for page links and titles)
 - [x] automatic rewrites (no more redirect pages, external rewrites are handled in next.config for Jotai)
 - [x] fetch configured targets via git
 - [x] wiki support
 - [x] vanilla frontmatter support (commented out, but the same syntax as MDX)
 - [x] vanilla markdown support (we rely on heavy MDX syntax, but some repos don't use it -- maybe magically insert some components)
 - [x] update strategy (ISR builds don't work at run-time w/o git and local storage, will have to configure git/deploy hooks)
 - ~~streamline syntax (if we do support MDX/vanilla magic, we have a bunch of redundant components like Demo, Sandbox, CodeSandbox)~~ -- we'll have to leave this to authors, especially since they can ship custom components in their repos

### CODE REVIEW

File: pages/[...slug].tsx [hunk 1]
```diff
@@ -53,14 +51,88 @@ export default function PostPage({ toc, source, allDocs, nav, frontMatter }) {
...
+
+const getPages = async (lib: keyof typeof settings) => {
+  const cached = cachedPages.get(lib)
+  if (cached) return cached
+
+  const { docs } = data.find(({ id }) => id === lib)
+  const { repo, dir, tag = 'master' } = docs
+
+  const { tree } = await fetch(
+    `https://api.github.com/repos/${repo}/git/trees/${tag}?recursive=1`
```
Comment (dai-shi) [inline pages/[...slug].tsx:71]: curious if it's technically possible fetch from github wiki pages.
ex. https://github.com/pmndrs/valtio.wiki.git

Comment (CodyJasonBennett) [reply pages/[...slug].tsx:64]: I'll have to give that a try. Cloning the repository (wikis can be cloned as repos) is a more robust way of doing this, although slower.

We'd have to get creative in caching by SHA or versioning and busting cache to trigger rebuilds.

**Comment by CodyJasonBennett** [comment]
Vercel doesn't like the formatting but this is what Valtio's wiki looks like.

image

**Comment by dai-shi** [comment]
wow, seems promising. having metadata in comments is totally acceptable.

**Comment by CodyJasonBennett** [comment]
Discovered some issues/pitfalls with the current system for markdown:
 - comments are left in the final eval/JSX code, so they have to be removed
 - vanilla markdown is evaluated as MDX, so stuff like `import ...` will break
 - escaped MDX syntax is evaluated as though it isn't escaped (\`jsx here\`)

**Comment by dai-shi** [comment]
Is there a workaround on wiki side? Not sure if I get what "isolated there" means.

**Comment by CodyJasonBennett** [comment]
I'll take a look. The issues are bugs/unsupported features with our current dependencies.

**Comment by CodyJasonBennett** [comment]
It seems that the use of space and `<br />` around paragraphs is tripping up our markdown parsers, so they're not parsed into components but plain text. Combined with eval, that will have the paragraph's contents evaluated as JSX.

We can possibly clean this up within the Wiki.

As-is from wiki:
image

Tidied:
image

File: pages/valtio/index.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import redirect from '../../utils/redirect'
+
+export default redirect('/valtio/home')
```
Comment (CodyJasonBennett) [inline pages/valtio/index.ts:3]: I think we can do away with this by redirecting paths to their first child, respecting `nav` order.

**Comment by CodyJasonBennett** [comment]
I'm hosting the wiki under [this repo](https://github.com/CodyJasonBennett/valtio), but you can see it deployed under [/valtio](https://website-6nhkftsfw-pmndrs.vercel.app/valtio/home) now.

I have frontmatter configured to preserve the nav order found on the wiki, and I've cleaned them up by replacing `<br />` with new lines and replaced h1s with h2s (`#` => `##`).

We might be able to automatically streamline docs with those changes.

File: pages/api/github/push.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  const modified = files.filter(
+    (file: string) => file.startsWith(repo.docs.dir) && MARKDOWN_REGEX.test(file)
+  )
+  if (!modified.length) return res.status(304)
+
+  // Update docs
+  await getDocs(repo.id, true)
+
+  return res.status(201).json({ message: `${modified.length} docs updated` })
+}
```
Comment (CodyJasonBennett) [inline pages/api/github/push.ts:41]: Maybe this isn't necessary if we can pull and check for changes via git. We can leverage [Next's ISR](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration) this way to rebuild on remote changes.

File: utils/mdx.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+/**
+ * Traverses a repo for matching docs, returning local paths.
+ */
+export const getPaths = async ({ repo, branch = 'main', dir = '' }: DocsConfig) => {
+  // Update repo if on disk, otherwise clone it
+  const repoDir = path.resolve(process.cwd(), `temp/${repo}-${branch}`)
+  if (!fs.existsSync(repoDir)) clone({ repo, dir, branch })
+  // else execSync(`git -C ${repoDir} pull`)
```
Comment (CodyJasonBennett) [inline utils/mdx.ts:87]: Need to figure out how to pull changes while not authenticated (read-only).

File: pages/[...slug].tsx [hunk 5]
```diff
@@ -54,21 +51,42 @@ export default function PostPage({ toc, source, allDocs, nav, frontMatter }) {
 }
 
 export const getStaticProps = async ({ params }) => {
-  const postFilePath = path.join(DOCS_PATH, `${path.join(...params.slug)}.mdx`)
-  const postData = fs.readFileSync(postFilePath)
-  const { content, data } = matter(postData)
+  const [lib] = params.slug
+  const docs = await getDocs(lib, false)
```
Comment (CodyJasonBennett) [inline pages/[...slug].tsx:55]: Leaving the invalidate param as `false` for now until we can find a workaround.

With invalidation enabled:
image

**Comment by CodyJasonBennett** [comment]
Just noticed that our build times are kinda slow on Vercel.




Is there any throttling I should know about? Not getting this on my machines.

File: pages/[...slug].tsx [hunk 6]
```diff
@@ -50,27 +50,35 @@ export default function PostPage({ allDocs, nav, toc, data, source }: PostPagePr
...
+export default function PostPage({ redirect, ...rest }: PostPageProps) {
+  const router = useRouter()
+
+  useLayoutEffect(() => {
+    if (redirect) return void router.push(redirect)
+  }, [redirect, router])
+
+  return redirect ? null : <Post {...rest} />
+}
+
```
Comment (CodyJasonBennett) [inline pages/[...slug].tsx:66]: Maybe we can serve the page specified as `redirect` instead? Ideally, we can change the URL as well without consequence. Looking into [*shallow routing*](https://nextjs.org/docs/routing/shallow-routing).

File: pages/api/update.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    // Check if docs were modified
+    const files = [...head_commit.added, ...head_commit.removed, ...head_commit.modified]
+    const changes = files.filter(
+      (file: string) =>
+        MARKDOWN_REGEX.test(file) && (!repo.docs.dir || file.startsWith(repo.docs.dir))
+    )
+    if (!changes.length) return res.status(304)
+
+    // Update docs
+    await fetch(process.env.VERCEL_DEPLOY_WEBHOOK, { method: 'POST' })
```
Comment (CodyJasonBennett) [inline pages/api/update.ts:56]: The webhook is currently aimed at this branch. Will need to point it to docs (prod) once merged.

**Pull request merged**